src/shared/sam/localLambdaRunner.ts

```diff
@@ -349,16 +319,63 @@ export async function invokeLambdaFunction(
                     (err as Error).message
                 )
             )
+
+            return false
         } finally {
             await remove(config.templatePath)
             telemetry.recordLambdaInvokeLocal({
-                lambdaPackageType: lambdaPackageType,
+                lambdaPackageType: isImageLambdaConfig(config) ? 'Image' : 'Zip',
                 result: invokeResult,
                 runtime: config.runtime as telemetry.Runtime,
                 debug: !config.noDebug,
                 version: samVersion,
             })
         }
+
+        return true
+    }
+}
+
+/**
+ * Prepares and invokes a lambda function via `sam local invoke`.
+ *
+ * @param ctx
+ * @param config
+ * @param onAfterBuild  Called after `SamCliBuildInvocation.execute()`
+ */
+export async function buildAndInvokeLambdaHandler(
+    ctx: ExtContext,
+    config: SamLaunchRequestArgs,
+    onAfterBuild: () => Promise<void>
+): Promise<SamLaunchRequestArgs> {
+    // Switch over to the output channel so the user has feedback that we're getting things ready
+    ctx.outputChannel.show(true)
+    if (!config.noDebug) {
+        const msg =
+            (config.invokeTarget.target === 'api' ? `API "${config.api?.path}", ` : '') +
+            `Lambda "${config.handlerName}"`
+        getLogger('channel').info(localize('AWS.output.sam.local.startDebug', 'Preparing to debug locally: {0}', msg))
+    } else {
+        getLogger('channel').info(
+            localize('AWS.output.sam.local.startRun', 'Preparing to run locally: {0}', config.handlerName)
+        )
+    }
+
+    const envVars = {
+        ...(config.awsCredentials ? asEnvironmentVariables(config.awsCredentials) : {}),
+        ...(config.aws?.region ? { AWS_DEFAULT_REGION: config.aws.region } : {}),
+    }
+
+    const timer = createInvokeTimer(ctx.settings)
+
+    if (!(await buildLambdaHandler(timer, envVars, config))) {
+        return config
+    }
```

Comment:
Just for reference: possibly the old logic intended to support cases like 39537851571ed where _invoke_ is still possible even though _build_ failed (e.g. if package.json is missing)--but this is a low priority. If that use-case becomes apparent for run/debug (as opposed to deploy), then we can think about supporting it on top of the structure of this PR.

Reply:
Yeah I thought about that. Logic could be added to 'buildLambdaHandler' to return true even if 'sam build' fails if we know the failure would not affect 'sam local invoke/api'.